Add `LARGE_BINARY` Attribute Type for External Storage of Large Fields

kunwp1 opened the pull request:
This PR introduces a new attribute type, `LARGE_BINARY`, to enable efficient handling of large binary fields by storing them externally in S3 rather than embedding them directly within tuples. It also includes various enhancements to support this new data type across the system.

---

### Motivation

Storing entire tuples in external storage adds unnecessary complexity and hinders direct access to small fields. Instead, we now store only individual large fields externally, which simplifies implementation and improves flexibility. Users can still access smaller fields directly in memory while large binary data is managed separately and efficiently.

---

### Design Overview

The new `large_binary` attribute type is distinct from the existing `binary` type:

- **`binary`**: Stores raw byte arrays directly in the tuple.
- **`large_binary`**: Stores a URI reference to an external binary object.

#### Lifecycle of `large_binary` fields

- **Creation**: Before emitting a tuple, the operator uploads the binary object to external S3 storage.
- **Transfer**: Tuples remain lightweight by storing only the URI.
- **Read**: Downstream operators use a utility API to resolve the URI and fetch the binary content.
- **Deletion**: Reference counting is used to manage deletion. When the count reaches zero, the binary object is deleted from S3.

---

### Implementation Details

- S3 buckets for storing large binary objects are created when the **computing unit master** is launched. The bucket name is defined in `storage.conf`.
- The `S3StorageClient` class was moved to `core/workflow-core` to be accessible from the computing unit master.
- `LARGE_BINARY` type is:
  - Stored as a **string** in Iceberg.
  - Distinguished from regular strings using a **magic prefix**: `"TEXERA_LARGE_BINARY:"` added to the attribute name.
  - If a string attribute uses the `"TEXERA_LARGE_BINARY:"` prefix, it will fail schema propagation, preventing accidental misinterpretation of types.
- A new PostgreSQL table stores the reference count for each URI for reference count tracking. 
- Transactions are used for concurrency control to ensure reference count consistency under concurrent insertions and deletions.
- The result export logic for cell data has been updated to resolve URIs back to actual binary content.
- **File uploads to S3** use **multipart upload** to ensure speed and reliability. Tested with a 3GB file, which took ~16 seconds depending on network conditions.
- The FileScan operator throws an error if a `BINARY` type is used with files larger than 2GB.

---

### Scope

This PR currently supports only the `FileScan` Java Native Operator.

---

### Migration Notice

After merging this PR, you must run the following SQL script to apply the necessary schema changes for reference count tracking:

```
core/scripts/sql/updates/08.sql
```

This script creates the required PostgreSQL table for storing reference counts of large binary URIs. Failing to run this script will result in runtime errors when handling `LARGE_BINARY` attributes.

--- 

### TODOs

- [ ] Add support for UDF operators  
- [ ] Investigate and update other Java Native Operators as needed  

<img width="1727" alt="Screenshot 2025-06-05 at 11 29 53 PM" src="https://github.com/user-attachments/assets/abd93f22-1de8-4a4f-8f13-c1f77da51fbd" />
<img width="1030" alt="Screenshot 2025-06-05 at 11 30 17 PM" src="https://github.com/user-attachments/assets/226aa6ad-7e79-43c4-8079-ab341d76d6b6" />
<img width="1728" alt="Screenshot 2025-06-05 at 11 30 06 PM" src="https://github.com/user-attachments/assets/cd588193-0d57-430b-a82a-53e939222716" />

bobbai00 commented on core/workflow-core/src/main/scala/edu/uci/ics/amber/util/IcebergUtil.scala:
```diff
@@ -47,12 +47,16 @@ import java.sql.Timestamp
 object IcebergUtil {
 
+  // Special prefix to identify LARGE_BINARY type in Iceberg
+  val LARGE_BINARY_PREFIX = "LARGE_BINARY_URI:"
```
rename to TEXERA_LARGE_BINARY_TYPE_ATTRIBUTE_NAME_PREFIX
change the value to "TEXERA_LARGE_BINARY:"

bobbai00 replied:
also move this const to Attribute.java, as it may be used by other components, not just iceberg

bobbai00 commented on core/workflow-operator/src/main/scala/edu/uci/ics/amber/operator/source/scan/FileScanSourceOpExec.scala:
```diff
@@ -81,6 +132,13 @@ class FileScanSourceOpExec private[scan] (
               fields.addOne(fileName)
             }
             fields.addOne(desc.attributeType match {
+              case FileAttributeType.LARGE_BINARY =>
```
please add a guard:
if the attribute name contains the TEXERA_LARGE_BINARY_PREFIX as its attribute name's prefix, then reject and prompt to users that the attribute name with that prefix is not allowed.

Need such guard for all the source operators and the Projection operator, and the Python UDF (because these operators can produce new attribute name).

kunwp1 replied:
I added a check in @Attribute.java. The schema propagation will fail if the string type attribute contains the prefix.

bobbai00 reviewed (commented):
Left some comments

kunwp1 commented:
@bobbai00 I've addressed your comments and also added support for concurrency control. Additionally, the UI has been updated to hide the URI from the user. The PR description has been updated accordingly. Here are the key highlights:

- A new PostgreSQL table is used to track the reference count for each URI.
- Transactions are employed to ensure concurrency control and maintain reference count consistency during concurrent insertions and deletions.

Please review the PR once more. Thanks.

kunwp1 commented:
I will close this PR and create a new PR based on this design #3787.
